r2dbc-mysql/src/main/java/io/asyncer/r2dbc/mysql/InitFlow.java

```diff
@@ -0,0 +1,747 @@
+import java.time.DateTimeException;
+import java.time.Duration;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * A message flow utility that can initializes the session of {@link Client}.
+ * <p>
+ * It should not use server-side prepared statements, because {@link PrepareCache} will be initialized after the session
+ * is initialized.
+ */
+final class InitFlow {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(InitFlow.class);
+
+    private static final ServerVersion MARIA_11_1_1 = ServerVersion.create(11, 1, 1, true);
+
+    private static final ServerVersion MYSQL_8_0_3 = ServerVersion.create(8, 0, 3);
+
+    private static final ServerVersion MYSQL_5_7_20 = ServerVersion.create(5, 7, 20);
+
+    private static final ServerVersion MYSQL_8 = ServerVersion.create(8, 0, 0);
+
+    private static final BiConsumer<ServerMessage, SynchronousSink<Boolean>> INIT_DB = (message, sink) -> {
+        if (message instanceof ErrorMessage) {
+            ErrorMessage msg = (ErrorMessage) message;
+            logger.debug("Use database failed: [{}] [{}] {}", msg.getCode(), msg.getSqlState(), msg.getMessage());
+            sink.next(false);
+            sink.complete();
+        } else if (message instanceof CompleteMessage && ((CompleteMessage) message).isDone()) {
+            sink.next(true);
+            sink.complete();
+        } else {
+            ReferenceCountUtil.safeRelease(message);
+        }
+    };
+
+    private static final BiConsumer<ServerMessage, SynchronousSink<Void>> INIT_DB_AFTER = (message, sink) -> {
+        if (message instanceof ErrorMessage) {
+            sink.error(((ErrorMessage) message).toException());
+        } else if (message instanceof CompleteMessage && ((CompleteMessage) message).isDone()) {
+            sink.complete();
+        } else {
+            ReferenceCountUtil.safeRelease(message);
+        }
+    };
+
+    /**
+     * Initializes handshake and login a {@link Client}.
+     *
+     * @param client                the {@link Client} to exchange messages with.
+     * @param sslMode               the {@link SslMode} defines SSL capability and behavior.
+     * @param database              the database that will be connected.
+     * @param user                  the user that will be login.
+     * @param password              the password of the {@code user}.
+     * @param compressionAlgorithms the list of compression algorithms.
+     * @param zstdCompressionLevel  the zstd compression level.
+     * @return a {@link Flux} that indicates the initialization is done, or an error if the initialization failed.
+     */
+    static Flux<Void> initHandshake(Client client, SslMode sslMode, String database, String user,
```

Comment:
maybe Mono?

Reply:
Yes, you are right, `Flux<Void>` should not exist.